File: renderer_test.ts
/**
 * Tests for ASCII art renderer
 */

import { assertEquals, assertStringIncludes } from "@std/assert";
import { parse } from "./parser.ts";
import { render } from "./renderer.ts";

Deno.test("render - Simple flowchart", () => {
  const input = `flowchart TD
    A[Hello] --> B[World]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  // Should contain both labels
  assertStringIncludes(output, "Hello");
  assertStringIncludes(output, "World");

  // Should contain box characters
  assertStringIncludes(output, "┌");
  assertStringIncludes(output, "┐");
  assertStringIncludes(output, "└");
  assertStringIncludes(output, "┘");
});

Deno.test("render - ASCII charset", () => {
  const input = `flowchart TD
    A[Test]`;

  const flowchart = parse(input);
  const output = render(flowchart, "ascii");

  // Should contain ASCII box characters
  assertStringIncludes(output, "+");
  assertStringIncludes(output, "-");
  assertStringIncludes(output, "|");
});

Deno.test("render - Unicode round charset", () => {
  const input = `flowchart TD
    A[Test]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode-round");

  // Should contain rounded corners
  assertStringIncludes(output, "╭");
  assertStringIncludes(output, "╮");
  assertStringIncludes(output, "╰");
  assertStringIncludes(output, "╯");
});

Deno.test("render - Unicode bold charset", () => {
  const input = `flowchart TD
    A[Test]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode-bold");

  // Should contain bold box characters
  assertStringIncludes(output, "┏");
  assertStringIncludes(output, "┓");
  assertStringIncludes(output, "┗");
  assertStringIncludes(output, "┛");
});

Deno.test("render - Unicode double charset", () => {
  const input = `flowchart TD
    A[Test]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode-double");

  // Should contain double box characters
  assertStringIncludes(output, "╔");
  assertStringIncludes(output, "╗");
  assertStringIncludes(output, "╚");
  assertStringIncludes(output, "╝");
});

Deno.test("render - LR direction", () => {
  const input = `flowchart LR
    A[Start] --> B[End]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  // Both nodes should be on same or similar lines
  const lines = output.split("\n");
  let startLine = -1;
  let endLine = -1;

  for (let i = 0; i < lines.length; i++) {
    if (lines[i].includes("Start")) startLine = i;
    if (lines[i].includes("End")) endLine = i;
  }

  // In LR mode, labels should be on the same row
  assertEquals(startLine, endLine);
});

Deno.test("render - Circle node shape", () => {
  const input = `flowchart TD
    A((Circle))`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  // Should contain double parentheses
  assertStringIncludes(output, "((");
  assertStringIncludes(output, "))");
  assertStringIncludes(output, "Circle");
});

Deno.test("render - Diamond node shape", () => {
  const input = `flowchart TD
    A{Decision}`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  assertStringIncludes(output, "Decision");
});

Deno.test("render - Hexagon node shape", () => {
  const input = `flowchart TD
    A{{Hexagon}}`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  assertStringIncludes(output, "{{");
  assertStringIncludes(output, "}}");
  assertStringIncludes(output, "Hexagon");
});

Deno.test("render - CJK labels with ambiguous width 1", () => {
  const input = `flowchart TD
    A[日本語]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode", 1);

  assertStringIncludes(output, "日本語");
});

Deno.test("render - CJK labels with ambiguous width 2", () => {
  const input = `flowchart TD
    A[日本語]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode", 2);

  assertStringIncludes(output, "日本語");
});

Deno.test("render - Empty flowchart returns empty string", () => {
  const input = `flowchart TD`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  assertEquals(output, "");
});

Deno.test("render - Multiple nodes", () => {
  const input = `flowchart TD
    A[First] --> B[Second]
    B --> C[Third]`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  assertStringIncludes(output, "First");
  assertStringIncludes(output, "Second");
  assertStringIncludes(output, "Third");
});

Deno.test("render - Stadium shape", () => {
  const input = `flowchart TD
    A([Stadium])`;

  const flowchart = parse(input);
  const output = render(flowchart, "unicode");

  assertStringIncludes(output, "([");
  assertStringIncludes(output, "])");
  assertStringIncludes(output, "Stadium");
});

File: renderer/renderer_test.go
package renderer

import (
	"strings"
	"testing"

	"github.com/kawaz/mermaid-aa/parser"
	"github.com/kawaz/mermaid-aa/types"
	"github.com/kawaz/mermaid-aa/width"
)

func TestRenderSimple(t *testing.T) {
	fc, _ := parser.ParseFlowchart("graph TD; A-->B")
	opts := &Options{
		Charset:        "ascii",
		AmbiguousWidth: 1,
	}
	result := RenderFlowchart(fc, opts)

	// Check that result is not empty
	if result == "" {
		t.Error("render result is empty")
	}

	// Check that it contains node labels
	if !strings.Contains(result, "A") {
		t.Error("result should contain node A")
	}
	if !strings.Contains(result, "B") {
		t.Error("result should contain node B")
	}
}

func TestRenderWithLabels(t *testing.T) {
	fc, _ := parser.ParseFlowchart("graph TD; A[Start]-->B[End]")
	opts := &Options{
		Charset:        "ascii",
		AmbiguousWidth: 1,
	}
	result := RenderFlowchart(fc, opts)

	if !strings.Contains(result, "Start") {
		t.Error("result should contain 'Start'")
	}
	if !strings.Contains(result, "End") {
		t.Error("result should contain 'End'")
	}
}

func TestRenderHorizontal(t *testing.T) {
	fc, _ := parser.ParseFlowchart("graph LR; A-->B-->C")
	opts := &Options{
		Charset:        "ascii",
		AmbiguousWidth: 1,
	}
	result := RenderFlowchart(fc, opts)

	// Check that result is not empty
	if result == "" {
		t.Error("render result is empty")
	}

	// In LR layout, nodes should be on the same or similar lines
	lines := strings.Split(result, "\n")
	if len(lines) < 1 {
		t.Error("result should have at least one line")
	}
}

func TestRenderCharsets(t *testing.T) {
	charsets := []string{"ascii", "unicode", "unicode-round", "unicode-bold", "unicode-double"}

	for _, cs := range charsets {
		t.Run(cs, func(t *testing.T) {
			fc, _ := parser.ParseFlowchart("graph TD; A[Test]-->B[Node]")
			opts := &Options{
				Charset:        cs,
				AmbiguousWidth: 1,
			}
			result := RenderFlowchart(fc, opts)

			if result == "" {
				t.Errorf("render result for charset %s is empty", cs)
			}
		})
	}
}

func TestRenderNodeShapes(t *testing.T) {
	shapes := []struct {
		syntax string
		shape  types.NodeShape
	}{
		{"A[text]", types.Rectangle},
		{"A(text)", types.Rounded},
		{"A([text])", types.Stadium},
		{"A{text}", types.Diamond},
		{"A{{text}}", types.Hexagon},
		{"A((text))", types.Circle},
		{"A[/text/]", types.Parallelogram},
		{`A[/text\]`, types.Trapezoid},
	}

	for _, tt := range shapes {
		t.Run(tt.shape.String(), func(t *testing.T) {
			fc, _ := parser.ParseFlowchart("graph TD; " + tt.syntax)
			opts := &Options{
				Charset:        "ascii",
				AmbiguousWidth: 1,
			}
			result := RenderFlowchart(fc, opts)

			if result == "" {
				t.Errorf("render result for shape %s is empty", tt.shape)
			}
			if !strings.Contains(result, "text") {
				t.Errorf("result should contain 'text' for shape %s", tt.shape)
			}
		})
	}
}

func TestRenderDirections(t *testing.T) {
	directions := []string{"TB", "TD", "BT", "LR", "RL"}

	for _, dir := range directions {
		t.Run(dir, func(t *testing.T) {
			fc, _ := parser.ParseFlowchart("graph " + dir + "; A-->B-->C")
			opts := &Options{
				Charset:        "ascii",
				AmbiguousWidth: 1,
			}
			result := RenderFlowchart(fc, opts)

			if result == "" {
				t.Errorf("render result for direction %s is empty", dir)
			}
		})
	}
}

func TestRenderUnicodeLabels(t *testing.T) {
	fc, _ := parser.ParseFlowchart("graph TD; A[日本語]-->B[テスト]")
	opts := &Options{
		Charset:        "unicode",
		AmbiguousWidth: 1,
	}
	result := RenderFlowchart(fc, opts)

	if !strings.Contains(result, "日本語") {
		t.Error("result should contain '日本語'")
	}
	if !strings.Contains(result, "テスト") {
		t.Error("result should contain 'テスト'")
	}
}

func TestRenderAmbiguousWidth(t *testing.T) {
	fc, _ := parser.ParseFlowchart("graph TD; A[Test]-->B[Node]")

	// Render with ambiguous width = 1
	opts1 := &Options{
		Charset:        "unicode",
		AmbiguousWidth: 1,
	}
	result1 := RenderFlowchart(fc, opts1)

	// Render with ambiguous width = 2
	opts2 := &Options{
		Charset:        "unicode",
		AmbiguousWidth: 2,
	}
	result2 := RenderFlowchart(fc, opts2)

	// Both should produce non-empty output
	if result1 == "" {
		t.Error("result with ambiguous=1 is empty")
	}
	if result2 == "" {
		t.Error("result with ambiguous=2 is empty")
	}
}

func TestCanvasOperations(t *testing.T) {
	calc := width.NewCalculator(1)
	canvas := NewCanvas(10, 5, calc)

	// Test Set and Get
	canvas.Set(0, 0, 'X')
	if canvas.Get(0, 0) != 'X' {
		t.Error("Set/Get failed")
	}

	// Test SetString
	canvas.SetString(2, 1, "Hi")
	if canvas.Get(2, 1) != 'H' {
		t.Error("SetString failed for 'H'")
	}
	if canvas.Get(3, 1) != 'i' {
		t.Error("SetString failed for 'i'")
	}

	// Test bounds
	canvas.Set(-1, 0, 'Y')  // should not panic
	canvas.Set(100, 0, 'Y') // should not panic

	// Test String output
	result := canvas.String()
	if !strings.Contains(result, "X") {
		t.Error("String output should contain 'X'")
	}
}
